common/TextDifferencer.java

```diff
@@ -0,0 +1,104 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.startup.common;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * An implementation of text difference based on the Longest Common Subsequence problem (A.K.A LCS).
+ */
+public class TextDifferencer {
+
+  /**
+   * Returns all the text differences between two given strings.
+   *
+   * @param first The first string.
+   * @param second The second string.
+   * @return A list which holds all the text differences.
+   */
+  public static List<CharDifference> getAllTextDifferences(String first, String second) {
+    return getDifferencesFromLCSMatrix(computeLCSMatrix(first, second), first, second);
+  }
+
+  /** Create an empty matrix based on the given dimentions. */
+  private static int[][] createEmptyLCSMatrix(int rowSize, int colSize) {
+    final int[][] lcsMatrix = new int[rowSize][];
+    for (int i = 0; i < lcsMatrix.length; ++i) {
+      lcsMatrix[i] = new int[colSize];
+    }
+    return lcsMatrix;
+  }
+
+  /**
+   * Precompute all the Longest Subsequence Matrix which holds the length of each common
+   * subsequence.
+   *
+   * @param first The first string out of two strings for precomputing a common subsequence.
+   * @param second The second string out of two strings for precomputing a common subsequence.
+   * @return A length of the common subsequence for the two strings.
+   */
+  private static int[][] computeLCSMatrix(String first, String second) {
+    final int[][] lcsMatrix = createEmptyLCSMatrix(first.length() + 1, second.length() + 1);
+    for (int i = 1; i < first.length() + 1; ++i) {
+      for (int j = 1; j < second.length() + 1; ++j) {
+        if (first.charAt(i - 1) == second.charAt(j - 1)) {
+          lcsMatrix[i][j] = lcsMatrix[i - 1][j - 1] + 1;
+        } else {
+          lcsMatrix[i][j] = Math.max(lcsMatrix[i][j - 1], lcsMatrix[i - 1][j]);
+        }
+      }
+    }
+    return lcsMatrix;
+  }
+
+  /**
+   * Compute all the character differences between two strings based on the precomputed Longest
+   * Common Subsequence matrix.
+   *
+   * @param lcsMatrix The precomputed longest subsequence matrix.
+   * @param first The first string of two strings to be compared with.
+   * @param second The second string of two strings to be compared with.
+   * @return A list of all the character differences
+   */
+  private static List<CharDifference> getDifferencesFromLCSMatrix(
+      int[][] lcsMatrix, String first, String second) {
```

Comment:
I think it would be more readable to pass arrays from `computeLCSMatrix()`, e.g: `char[] first` (`first.toCharArray()`)

Reply:
@oferb I think it is much more personal preference, but I don't have strong feelings about it, so Done.